src/transform/src/fusion/reduce.rs

```diff
@@ -0,0 +1,123 @@
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! Fuses reduce operators with parent operators if possible.
+
+use crate::TransformArgs;
+use expr::{AggregateExpr, AggregateFunc, MirRelationExpr, MirScalarExpr};
+
+/// Fuses reduce operators with parent operators if possible.
+#[derive(Debug)]
+pub struct Reduce;
+
+impl crate::Transform for Reduce {
+    fn transform(
+        &self,
+        relation: &mut MirRelationExpr,
+        _: TransformArgs,
+    ) -> Result<(), crate::TransformError> {
+        relation.visit_mut_pre(&mut |e| {
+            self.action(e);
+        });
+        Ok(())
+    }
+}
+
+impl Reduce {
+    /// Fuses reduce operators with parent operators if possible.
+    pub fn action(&self, relation: &mut MirRelationExpr) {
+        if let MirRelationExpr::Reduce {
+            input,
+            group_key,
+            aggregates,
+            monotonic: _,
+            expected_group_size: _,
+        } = relation
+        {
+            if let MirRelationExpr::Reduce {
+                input: inner_input,
+                group_key: inner_group_key,
+                aggregates: inner_aggregates,
+                monotonic: _,
+                expected_group_size: _,
+            } = &mut **input
+            {
+                // Do nothing if outer key is not a subset of inner key
+                if !group_key
+                    .iter()
+                    .all(|e| matches!(e, MirScalarExpr::Column(_)) && inner_group_key.contains(e))
+                {
+                    return;
+                }
+
+                if aggregates.is_empty() && inner_aggregates.is_empty() {
+                    // Replace inner reduce with map + project (no grouping)
+                    let mut outputs = vec![];
+                    let mut scalars = vec![];
+                    for e in inner_group_key {
+                        if let MirScalarExpr::Column(i) = e {
+                            outputs.push(*i);
+                        } else {
+                            scalars.push(e.clone());
+                        }
+                    }
+
+                    let arity = inner_input.arity();
+                    for i in 0..scalars.len() {
+                        outputs.push(arity + i);
+                    }
+
+                    **input = inner_input.take_dangerous().map(scalars).project(outputs);
+                } else if aggregates.len() == 1 && inner_aggregates.len() == 1 {
+                    // TODO: This can be more general than just len() == 1
+                    // Drop inner reduce and rewrite outer reduce if possible
+                    if let AggregateExpr {
+                        func: AggregateFunc::Count,
+                        expr,
+                        distinct: _,
+                    } = inner_aggregates.first().unwrap()
+                    {
+                        if inner_group_key.contains(expr) {
```

Comment:
Note that this is just covering the case where this COUNT aggregation can be simply replaced with 1, since it always evaluates to that (sincer the inner expression is part of the grouping key).

Reply:
Makes sense. Note that this was removed to limit the scope to cases where the outer reduction doesn't depend at all on the result of any aggregation performed by the inner reduction.